contrib/babelfishpg_common/src/datetime.c

```diff
@@ -663,6 +666,454 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+	char *pattern, *temp_field;
+	int len_str = strlen(str);
+	char *modified_str = (char*) palloc(len_str + 1);
+	int j = 0, i = 0;
+
+	/* If sql_dialect is not an TSQL then return false */
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return false;
+
+	/* Throw a common error message while casting to time datatype */
+	#define TIME_IN_ERROR()	\
+		ereport(ERROR,	\
+				(errcode(ERRCODE_INVALID_DATETIME_FORMAT),	\
+				 errmsg("Conversion failed when converting date and/or time from character string.")));	\
+
+	/*
+	 * The below logic removes the space present before and after of ':', '/',
+	 * '.', '-' and also removes the multiple whitespaces to single whitespace.
+	 * For example converts the input from "    1  :     1" to " 1:1"
+	 */
+	while (i < len_str)
+	{
+		if (str[i] == ' ' && (i == (len_str - 1) || str[i+1] == ' ' || str[i+1] == ':' || str[i+1] == '/' || str[i+1] == '.' || str[i+1] == '-'))
+		{
+			i++;
+			continue;
+		}
+		else if (str[i] == ':' || str[i] == '/' || str[i] == '.' || str[i] == '-')
+		{
+			modified_str[j++] = str[i++];
+			while (i < len_str && str[i] == ' ')
+				i++;
+		}
+		else
+			modified_str[j++] = str[i++];
+	}
+	modified_str[j] = '\0';
+
+	dterr = ParseDateTime(modified_str, workbuf, sizeof(workbuf),
+			field, ftype, MAXDATEFIELDS, &nf);
+
+	/*
+	 * If there are no errors while parsing the modified_str then
+	 * validate the fields based on their types.
+	 */
+	if (dterr == 0)
+	{
+		if (nf >= 3)
+		{
+			/*
+			 * If the total no.of fields is >=3 then there is an possiblity that
+			 * the modified_str contains text month and can be of any format
+			 * {"DD MON YYYY", "DD YYYY MON", "MON DD YYYY", "MON YYYY DD",
+			 * "YYYY MON DD", "YYYY DD MON"} then convert the fields to an valid
+			 * format of {"DD-MON-YYYY", "YYYY-MON-DD"} and change type to "DTK_DATE"
+			 */
+			if (containsInTextMonthFormat(ftype, field))
+			{
+				/*
+				 * The below logic will append the field[1], field[2] to the field[0]
+				 * and change the type of field[0] to "DTK_DATE".
+				 * For example if input is "2000 nov 23" will be converted
+				 * to "2000-nov-23".
+				 */
+				res = makeStringInfo();
+				appendStringInfo(res, "%s-%s-%s", field[0], field[1], field[2]);
+				field[0] = res->data;
+				ftype[0] = DTK_DATE;
+				pfree(res);
+
+				/*
+				 * Since the first 3 fields appended to 1st field, skip the
+				 * attached fields i.e., field[1]/[2], ftype[1]/[2]
+				 */
+				nf = nf - 2;
+				for(int i = 1; i < nf; i++)
+				{
+					field[i] = field[i+2];
+					ftype[i] = ftype[i+2];
+				}
+			}
+		}
+
+		/*
+		 * Iterate through each fields and do pre-validation of that field,
+		 * throw error if the field is in invalid format.
+		 */
+		for (int i = 0; i < nf; i++)
+		{
+			int len;
+			char* different_date_formats[] = {"/", ".", "-"};
+			temp_field = pstrdup(field[i]);
+			len = strlen(temp_field);
+
+			switch (ftype[i])
+			{
+				case DTK_NUMBER:
+					/*
+					 * If there is an number field present in the field
```

Comment:
1. nit: an number -> a number
2. Instead of saying "number field present in the field", we can say "number present in the field"

Reply:
Will fix in next revision